### DISCUSSION TITLE
Image.write_to_binary/1 seems slow?

### DISCUSSION THREAD

**Discussion by kipcole9** [akash-akya/vix#142] [Q&A]
@akash-akya, I'm trying to support @mat-hek with consuming and producing [YUV images](https://en.wikipedia.org/wiki/YCbCr). Whilst [functionally complete](https://github.com/elixir-image/image/blob/yuv/lib/image/yuv.ex) I am currently not able to meet his [performance expectations](https://github.com/elixir-image/image/discussions/121#discussioncomment-8068315).

When [benchmarking](https://github.com/elixir-image/image/blob/yuv/bench/write_to_binary.exs), the primary culprit seems to be `Vix.Vips.Image.write_to_binary/1` which, for a 1920x1080 RGB image takes about 10.8ms. Indeed even when trying to write just a single band of the image it takes the same amount of time.

Given that the test image is memory resident, and there is no operations pipeline, 10+ ms seems much higher than I would have expected.

Do you have any explanation or ideas?  Or are my expectations incorrect? I wanted to check with you before I ask John's advice as well.

As always, many thanks.

**Comment by akash-akya** [comment]
@kipcole9, `write_to_binary` has to decompress/unpack every pixel from the whatever image format source is in and write memory, right? It is not a noop. So it kinda makes sense that writing a single band takes the same time as writing all bands.

> Given that the test image is memory resident

Not sure what you mean, the image won't be in memory when you open right? but it is loaded when the write_to_binary operation is called.

That said I am not sure if 10+ms is higher or lower. Do you have some baseline expectations, perhaps from other libraries.

I tried to check if it is coming from the libvips or vix,
```
vips_image.c:864	(fn "nif_image_write_to_binary")  - vips_image_write_to_memory took: 59.540000ms
12:47:57.942 [debug] Vix write_to_binary took: 59.659ms
```
So It seems libvips call itself takes the whole time and not Vix.

**Reply by akash-akya** [reply]
Let me know If I miss something. I haven't checked the complete discussion on YUV support, will do that this evening 😅

**Reply by mat-hek** [reply]
>  I haven't checked the complete discussion on YUV support, will do that this evening

Thanks!

> Do you have some baseline expectations, perhaps from other libraries

I can try to prepare some comparisons, but generally, it's slower than video encoding to h264 on CPU using FFmpeg. I know it's a different story, and the encoder is well-optimized, but the encoding process is also much more complex. Another thing is that JPEG -> overlay image -> JPEG takes way slower than taking a YUV image and calling `write_to_binary`, while JPEG also converts to YUV underneath:

```elixir
Mix.install([:vix, {:image, github: "elixir-image/image", branch: "yuv"}, :benchee])

alias Vix.Vips.Image, as: Vimage

overlay = Image.open!("membrane.png")
yuv = File.read!("image.yuv")
{:ok, rgb} = Image.YUV.new_from_binary(yuv, 1920, 1080, :C420)

jpeg = Image.open!("image.jpeg")

Benchee.run(%{
  yuv: fn ->
    Vimage.write_to_binary(rgb)
  end,
  jpeg: fn ->
    {:ok, img} = Image.compose(jpeg, overlay)
    Vimage.write_to_buffer(img, ".jpg")
  end
})
```
```
Name           ips        average  deviation         median         99th %
jpeg         67.85       14.74 ms    ±32.42%       13.68 ms       43.02 ms
yuv          39.25       25.48 ms     ±4.67%       25.14 ms       30.35 ms

Comparison:
jpeg         67.85
yuv          39.25 - 1.73x slower +10.74 ms
```

[images.zip](https://github.com/akash-akya/vix/files/13885225/images.zip)

 I generally need to do the conversion from YUV to RGB and back in at most ~25ms for a 1280x720 image and hope that's possible with Vips ;)

**Comment by kipcole9** [comment]
@akash-akya thanks for kick-starting my brain. I added a one more test to the [benchmark](https://github.com/elixir-image/image/blob/yuv/bench/write_to_binary.exs) that  writes a generated image and the results reflect what you said. The writing to binary is only 1.33ms. So It's probably them the `Operation.recomb/2` thats much ofl the time. Do you know of a way to "force" the operation pipeline to execute?

```
Name                                 ips        average  deviation         median         99th %
Write a generated image           750.30        1.33 ms    ±38.60%        1.30 ms        2.92 ms
Write full image to binary         96.79       10.33 ms     ±7.78%       10.57 ms       11.87 ms
Write one band to binary           94.98       10.53 ms     ±7.84%       10.80 ms       12.00 ms

Comparison:
Write a generated image           750.30
Write full image to binary         96.79 - 7.75x slower +9.00 ms
Write one band to binary           94.98 - 7.90x slower +9.20 ms
```

**Reply by akash-akya** [reply]
If recomb is the bottleneck then forcing pipeline does not solve the issue right? 
If you are asking this for testing, then one hacky way might be to just call any `write_xxx` first to cache the image. And hopefully when we call a `write_xxx` second time, that time we use cached image and thus only paying for `write_xxx`. It is a bit hard to prove that exactly this happen though.

Maybe we have to use low level primitive libvips operation to get better perf. Let me check the yuv code once.

Btw you can use `vipsprofile` to get some more info about concurrency. Some details about it [here](https://libvips.blogspot.com/2013/11/profiling-libvips.html)

**Comment by kipcole9** [comment]
Good suggestion. I updated the [benchmark](https://github.com/elixir-image/image/blob/yuv/bench/write_to_binary.exs) to do an `Image.write_to_binary/1` before the benchmark loop. Unfortunately no change to performance:

```
Name                                 ips        average  deviation         median         99th %
Write a generated image           821.62        1.22 ms    ±42.71%        1.15 ms        2.81 ms
Write one band to binary           98.55       10.15 ms     ±8.22%       10.42 ms       11.67 ms
Write full image to binary         97.22       10.29 ms     ±7.51%       10.40 ms       12.01 ms
```

I will take a look about `vipsprofile` now - I wasn't aware of that!

**Reply by akash-akya** [reply]
Interesting. I have to find explanation for this behavior too :)

**Reply by kipcole9** [reply]
I calculated the time for `wirite_to_binary/1` by first executing my pipeline (`Image.YUV.to_yuv/2`) with the `recomb/2` and subsample/3` included, and then without. The different is the pipeline performance (about 3ms) and the remainder is the `write_to_binary/1` (about 10ms). No matter what I do the `write_to_binary/1` is about 10ms.

**Reply by kipcole9** [reply]
And of course `Vix.Vips.Image.write_to_binary/1` is just a NIF call so I know it's not your code. But I also know you understand how `libvips` works much better than I do!

**Comment by akash-akya** [comment]
@kipcole9 @mat-hek I checked YUV and played around a bit. It seems resizing the image is taking significant time. I got 33% gain by using `:VIPS_KERNEL_NEAREST` kernel while resizing. 
Other things I tried:
* using `Operation.linear!(image_444, [1], [-16, -128, -128], uchar: true)` instead of `image_444 - @yuv_to_rgb_offsets` but only resulted in minor gain. 
* resizing `uv` together by `bandjoin`ing first and then resizing (that way we resize once), but that too didn't yield significant improvement

Other than that, I couldn't find anything much. 

I think we should find a way to optimize scaling an image by `2`. I'll dig more on this later. 

Found this discussion in libvips: https://github.com/libvips/libvips/issues/2560

**Reply by akash-akya** [reply]
Another thing I noticed is that the execution time varies a lot across runs for same input. I see almost 7ms to 8ms variance, which seems significant?

**Reply by mat-hek** [reply]
> I see almost 7ms to 8ms variance, which seems significant?

Maybe it depends on the current load of your machine? I noticed that the conversion takes almost all of my CPU

> I got 33% gain by using :VIPS_KERNEL_NEAREST kernel while resizing.

That's great, I see that too!

> I think we should find a way to optimize scaling an image by 2. I'll dig more on this later.

Scaling takes a significant amount of time, but even without it, I get around 58ms for 720p image on my machine, which is over twice too slow. However, I noticed that replacing

```elixir
[
  Vimage.write_to_binary(yuv[0]),
  Vimage.write_to_binary(yuv[1]),
  Vimage.write_to_binary(yuv[2])
]
```

with

```elixir
Vimage.write_to_binary(yuv)
```

gives 3x overal speedup - 18ms instead of 58...

**Reply by kipcole9** [reply]
Yes, I see the same. But of course that produces a 4:4:4 YUV image, not a 4:2:0 image.  `Operation.subsample/3` seems to be the fastest way to subsample U and V but there's no escaping the need to subsample - unless I'm mistaken something?

Thats why I suggested the issue is `write_to_binary/1` performance as best I can establish.

**Reply by kipcole9** [reply]
> Found this discussion in libvips: https://github.com/libvips/libvips/issues/2560

Yes, thats where I got the reference implementation for `to_rgb` from for `image`. So far there isn't any performance issue with that conversion.

The issue is with the conversion back to YUV.  And there I use `Operation.subsample/3` rather than `Operation.resize`.  It does a block resize which is more efficient when reducing the image size.

**Reply by akash-akya** [reply]
Noted. Interesting finds, thanks for sharing! I have to spend more time.

@mat-hek can you share the 720p image you mentioned?

**Reply by kipcole9** [reply]
There is a test image from @mat-hek in the `image` repo at `test/support/images/image.yuv` although that is a 1080p image.   

The benchmark script in `bench/write_to_binary.exs` demonstrates what I see as the issue (which I appreciate isn't aligned yet with your observations).

**Reply by akash-akya** [reply]
@kipcole9 I am using your `yuv` branch to test. Which bench script are you testing? 
I am checking this [script](https://github.com/elixir-image/image/blob/yuv/bench/write_to_binary.exs) and it calls `to_rgb` which uses `new_scaled_image` so it uses `resize` not subsample right?

**Reply by kipcole9** [reply]
Akash, `new_scaled_image` is called in the conversion from YUV to RGB. So far that isn't the performance bottleneck. (Perhaps it will be later).

The issue is in RGB to YUV (as a binary). Therefore the script `yuv_to_rgb.exs` is the best illustration. It shows the times for conversion in each direction.

**Reply by kipcole9** [reply]
Here's what I see on my machine for that script:
```
Name                                                           ips        average  deviation         median         99th %
Converting YUV 4:2:0 binary in BT601 to an RGB image        1.42 K        0.70 ms    ±48.97%        0.68 ms        2.33 ms
Converting an RGB image to YUV 4:2:0 binary in BT601      0.0691 K       14.48 ms     ±8.09%       14.79 ms       16.63 ms

Comparison:
Converting YUV 4:2:0 binary in BT601 to an RGB image        1.42 K
Converting an RGB image to YUV 4:2:0 binary in BT601      0.0691 K - 20.57x slower +13.77 ms
```
Note that all image data starts as a raw binary - so there is no image decoding, decompression or headers. It's all just binary. So I *thing* its an apples-to-apples comparison but I may have missed something.

**Reply by mat-hek** [reply]
Sorry for the confusion, I generally test on both 1080 and 720, it seems that the latter is usually 2-2.5x faster. Here's the 720p image: [image720.yuv.zip](https://github.com/akash-akya/vix/files/13901443/image720.yuv.zip)

**Reply by kipcole9** [reply]
And .... thats showing 14.5ms for RGB to YUV. I swear the same test yesterday was showing 42ms !!!!!!

@mat-hek - if its reproducible at 14.5ms for your 1080p image, does that get us close?

**Reply by mat-hek** [reply]
> if its reproducible at 14.5ms for your 1080p image, does that get us close?

Yeah, seems very close ;)

**Reply by akash-akya** [reply]
@kipcole9 `bench/yuv_to_rgb.exs` script is slightly misleading though, `Image.YUV.new_from_binary` does not actually do anything on libvips side, since everything is just put to pipeline without a sink to force it. And `write_to_binary` runs the whole pipeline, including the pending yuv→rgb conversion related logic along with rgb→yuv logic.

Btw, we can force the pipeline by using any arithmetic operation such as `Vix.Vips.Operation.avg!(image)`

If I change the script to force the pipeline using avg then I get different results

```elixir
raw_yuv_file = Path.expand("./test/support/images/image.yuv")
{:ok, binary} = File.read(raw_yuv_file)
{:ok, image} = Image.YUV.new_from_binary(binary, 1920, 1080, :C420, :bt601)

Benchee.run(
  %{
    "Converting YUV 4:2:0 binary in BT601 to an RGB image" => fn ->
      {:ok, i} = Image.YUV.new_from_binary(binary, 1920, 1080, :C420, :bt601)
      Vix.Vips.Operation.avg!(i)
    end,

    "Converting an RGB image to YUV 4:2:0 binary in BT601" => fn ->
      {:ok, _} = Image.YUV.write_to_binary(image, :C420, :bt601)
    end
  },
  time: 20,
  memory_time: 2
)
```

**Reply by kipcole9** [reply]
Thanks you Akash, good callout and yes, its means the result of 11ms (on my system) for `new_from_binary` clearly is more in the expected range (compared to sub-millisecond).  But interestingly the work to build the RGB image involves a `Image.recomb`, requires interleaving the data into `rgbrgbrgb.....` layout. And yet it is 11ms. So for me it still doesn't explain why `write_to_binary` is so compartively "slow".  I know its not a `vix` issue but write to binary should be almost nothing more than a `memcpy`, no?

**Reply by akash-akya** [reply]
Yeah. Hard to get clear sense of what is taking time, perhaps difficulty in debugging is the cost of making things lazy 😅

**Comment by kipcole9** [comment]
I went back to basics and just focused on comparing `write_to_binary/1` for the same image in different formats (using the 1080p image in TIF, JPG and YUV formats). And then one example of writing just one band of that image. 

TLDR; the performance constraint isn't `Operation.write_to_binary/1`.

### Performance results
```
Name                                     ips        average  deviation         median         99th %
Write a generated black image         1.31 K      764.17 μs    ±42.49%      711.71 μs     2076.27 μs
Write full TIF image to binary        1.25 K      798.48 μs    ±36.33%      786.42 μs     1993.53 μs
Write full JPG image to binary        1.25 K      801.85 μs    ±36.66%      787.63 μs     2014.34 μs
Write one band to binary              1.00 K      995.49 μs    ±34.97%      984.10 μs     2256.65 μs
Write full YUV image to binary       0.100 K     9997.92 μs     ±7.68%    10216.69 μs    11486.27 μs
```

### Observations

* A generated image, a decoded JPG image and a TIF image (which is raw RGB data inside a container) all show sub-millisecond performance - which is what I would expect.
* The YUV image - which includes all of the conversion code - takes 10ms (as we have been seeing). The difference can only be the conversion process (`Operation.recomb` etc etc).

### Conclusion

My current conclusion is that it is the conversion process that is taking the time. This is similar to the findings in the libvips discussion (albeit that was focused more on the conversion TO rgb.

@mat-hek I'm not seeing a path to delivering 30fps processing for 1080p, and probably the same for 720p, images.

Unless there are any other insights or ideas I can try, I propose closing this discussion.

**Reply by akash-akya** [reply]
I guess we could look into other libraries or yuv decoder to find optimization paths. That would take time though

**Comment by kipcole9** [comment]
@akash-akya, yes, I'd say [libyuv from the Google Chrome project](https://chromium.googlesource.com/libyuv/libyuv/) would be the best place to start for a way to get native performance. But ..... let's just say the documentation is "sparse". :-). And as you know, my NIF-fu is not good enough to tackle that with any confidence.

Closing the discussion for now, thanks for your patience @akash-akya, and sorry for not finding a path to the performance you need @mat-hek.

**Reply by kipcole9** [reply]
Oh, also turns out that `libavif` vendors `libyuv` and wraps it with a more approachable API.  So that might be a better angle.

@mat-hek, I'm up for the challenge adding a NIF interface to `libavif` - but I can't commit a timeline. And I can't commit I have enough skills either.

**Comment by mat-hek** [comment]
Ok, I tried an approach with overlaying separate planes and it seems it works and it's fast! 13ms for 720p, 18ms for 1080p on my machine. The only problem is that it doesn't consider the alpha channel.

```elixir
Mix.install([:vix, {:image, github: "elixir-image/image", branch: "yuv"}, :benchee])

defmodule Overlay do
  alias Vix.Vips.Image, as: Vimage

  def run() do
    overlay = Image.open!("membrane.png")
    yuv = File.read!("image.yuv")
    {:ok, overlay_yuv} = Image.YUV.write_to_binary(overlay, :C420)
    {overlay_y, overlay_u, overlay_v} = open_planes(overlay_yuv, 400, 400)

    # Benchee.run(%{
    #   yuv: fn ->
    {image_y, image_u, image_v} = open_planes(yuv, 1920, 1080)
    composed_y = compose_to_binary(image_y, overlay_y)
    composed_u = compose_to_binary(image_u, overlay_u)
    composed_v = compose_to_binary(image_v, overlay_v)
    composed = composed_y <> composed_u <> composed_v
    #   end
    # })

    {:ok, output} = Image.YUV.new_from_binary(composed, 1920, 1080, :C420)
    Image.write!(output, "output.jpeg")
  end

  defp open_planes(yuv, width, height) do
    half_width = div(width, 2)
    half_height = div(height, 2)
    y_size = width * height
    uv_size = half_width * half_height
    <<y::binary-size(y_size), u::binary-size(uv_size), v::binary-size(uv_size)>> = yuv
    {:ok, y} = Vimage.new_from_binary(y, width, height, 1, :VIPS_FORMAT_UCHAR)
    {:ok, u} = Vimage.new_from_binary(u, half_width, half_height, 1, :VIPS_FORMAT_UCHAR)
    {:ok, v} = Vimage.new_from_binary(v, half_width, half_height, 1, :VIPS_FORMAT_UCHAR)
    {y, u, v}
  end

  defp compose_to_binary(image, overlay) do
    {:ok, binary} = Image.compose!(image, overlay) |> Image.flatten!() |> Vimage.write_to_binary()
    binary
  end
end

Overlay.run()
```

And it outputs

![output](https://github.com/akash-akya/vix/assets/16493463/fb415acf-81ea-4194-8c2c-e2ffa5c38ec9)

Any ideas on how I can make it consider the alpha channel of the overlay? Even binary alpha (either fully transparent or no transparency) could be enough.

**Reply by kipcole9** [reply]
That's really interesting but I don't know how to apply to alpha programmatically.`Image.flatten/1` is a noop in your code since you're operating on single-band images.  In addition, you're decomposing your overlay image into y, u and v - but no alpha. So basically there is no alpha processing in your pipeline.  Sorry I don't have any clever ideas - although it is interesting that your code also points at the conversion from RGB to YUV is the primary performance hog.

I've been doing more research and it seems `libheif` might be the best library to use to accelerate RGB<->YUV conversions. In part because it's already built into most configurations of libvips and in part because it seems to have a sane API.  I'll be doing some research over the weekend.  Hopefully I can get to a `Image.YUV.write_to_buffer/2` that becomes a NIF call to `libheif`.  Then it would also have more control over subsampling, bit depth and colourspace.

**Reply by kipcole9** [reply]
I'll continue this thread over in the `image` library.

**Reply by akash-akya** [reply]
@mat-hek that's a neat trick! Adjusting it to use alpha channel can be done by simply joining alpha band we get from the overlay png image.

```elixir
defmodule Overlay do
  alias Vix.Vips.Image, as: Vimage
  alias Vix.Vips.Operation

  def run() do
    overlay = Image.open!("membrane.png")
    yuv = File.read!("image.yuv")

    # extract alpha channel from the overlay image
    alpha = overlay[3]
    downsized_alpha = Operation.subsample!(alpha, 2, 2)
    # Kino.render(alpha)

    {:ok, overlay_yuv} = Image.YUV.write_to_binary(overlay, :C420)
    {overlay_y, overlay_u, overlay_v} = open_planes(overlay_yuv, 400, 400)

    # add alpha channel
    overlay_y = Operation.bandjoin!([overlay_y, alpha])
    overlay_u = Operation.bandjoin!([overlay_u, downsized_alpha])
    overlay_v = Operation.bandjoin!([overlay_v, downsized_alpha])

    # Benchee.run(%{
    #   yuv: fn ->
        {image_y, image_u, image_v} = open_planes(yuv, 1920, 1080)
        composed_y = compose_to_binary(image_y, overlay_y)
        composed_u = compose_to_binary(image_u, overlay_u)
        composed_v = compose_to_binary(image_v, overlay_v)
        composed = composed_y <> composed_u <> composed_v
    #   end
    # })

    {:ok, output} = Image.YUV.new_from_binary(composed, 1920, 1080, :C420)
    Image.write!(output, "output.jpeg")
  end

  defp open_planes(yuv, width, height) do
    half_width = div(width, 2)
    half_height = div(height, 2)
    y_size = width * height
    uv_size = half_width * half_height
    <<y::binary-size(y_size), u::binary-size(uv_size), v::binary-size(uv_size)>> = yuv
    {:ok, y} = Vimage.new_from_binary(y, width, height, 1, :VIPS_FORMAT_UCHAR)
    {:ok, u} = Vimage.new_from_binary(u, half_width, half_height, 1, :VIPS_FORMAT_UCHAR)
    {:ok, v} = Vimage.new_from_binary(v, half_width, half_height, 1, :VIPS_FORMAT_UCHAR)
    {y, u, v}
  end

  defp compose_to_binary(image, overlay) do
    composed = Image.compose!(image, overlay)
    {:ok, binary} = Vimage.write_to_binary(composed[0])
    binary
  end
end

Overlay.run()
```

**Reply by akash-akya** [reply]
Also, I tried another approach using `ifthenelse` instead of compose to overlay images, and it is **lot** faster. Saw `5.5ms` for 1080p image instead of `17~18ms`.  But the issue is, currently it overlays the image in the top left corner and I don't know how to move it to the desired location without sacrificing the performance. 

If you want to play with that: 

```elixir
defmodule Overlay do
  alias Vix.Vips.Image, as: Vimage
  alias Vix.Vips.Operation

  def run() do
    overlay = Image.open!("membrane.png")
    yuv = File.read!("image.yuv")

    alpha = overlay[3]
    downsized_alpha = Operation.subsample!(alpha, 2, 2)
    # Kino.render(alpha)
    
    {:ok, overlay_yuv} = Image.YUV.write_to_binary(overlay, :C420)
    {overlay_y, overlay_u, overlay_v} = open_planes(overlay_yuv, 400, 400)

    # Benchee.run(%{
    #   yuv: fn ->
        {image_y, image_u, image_v} = open_planes(yuv, 1920, 1080)
        composed_y = compose_to_binary(image_y, overlay_y, alpha)
        composed_u = compose_to_binary(image_u, overlay_u, downsized_alpha)
        composed_v = compose_to_binary(image_v, overlay_v, downsized_alpha)
        composed = composed_y <> composed_u <> composed_v
    #   end
    # })

    {:ok, output} = Image.YUV.new_from_binary(composed, 1920, 1080, :C420)
    Image.write!(output, "output.jpeg")
  end

  defp open_planes(yuv, width, height) do
    half_width = div(width, 2)
    half_height = div(height, 2)
    y_size = width * height
    uv_size = half_width * half_height
    <<y::binary-size(y_size), u::binary-size(uv_size), v::binary-size(uv_size)>> = yuv
    {:ok, y} = Vimage.new_from_binary(y, width, height, 1, :VIPS_FORMAT_UCHAR)
    {:ok, u} = Vimage.new_from_binary(u, half_width, half_height, 1, :VIPS_FORMAT_UCHAR)
    {:ok, v} = Vimage.new_from_binary(v, half_width, half_height, 1, :VIPS_FORMAT_UCHAR)
    {y, u, v}
  end

  defp compose_to_binary(image, overlay, alpha) do
    {:ok, binary} =
      Operation.ifthenelse!(alpha, overlay, image)
      |> Vimage.write_to_binary()

    binary
  end
end

Overlay.run()
```
![img](https://github.com/akash-akya/vix/assets/9396022/68bf2075-913d-4292-b887-488ba76127af)

**Reply by akash-akya** [reply]
Okay, I found a middle ground. You can just create a new overlay image with offset to match base image. And this approach takes `10ms`. Basically, the performance is directly proportional to the overlay image size. Performs best if the overlay is at the top-left and worse if it is at bottom-right.

```elixir
defmodule Overlay do
  alias Vix.Vips.Image, as: Vimage
  alias Vix.Vips.Operation

  def run() do
    overlay = Image.open!("membrane.png")

    resized_overlay_path =
      resized_and_positioned_overlay(overlay,
        # we try to create a smallest possible overlay image with transparent offset from 
        # the top-left corner of base image
        width: div(1920, 2) + div(Vimage.width(overlay), 2),
        height: div(1080, 2) + div(Vimage.height(overlay), 2),
        # position of the overlay within the new transparent base
        x: div(1920, 2) - div(Vimage.width(overlay), 2),
        y: div(1080, 2) - div(Vimage.height(overlay), 2)
      )

    overlay = Image.open!(resized_overlay_path)
    yuv = File.read!("image.yuv")

    Kino.render(overlay)
    alpha = overlay[3]
    downsized_alpha = Operation.subsample!(alpha, 2, 2)

    # Kino.render(alpha)
    {:ok, overlay_yuv} = Image.YUV.write_to_binary(overlay, :C420)

    {overlay_y, overlay_u, overlay_v} =
      open_planes(overlay_yuv, Vimage.width(overlay), Vimage.height(overlay))

    # Benchee.run(%{
    #   yuv: fn ->
        {image_y, image_u, image_v} = open_planes(yuv, 1920, 1080)
        composed_y = compose_to_binary(image_y, overlay_y, alpha)
        composed_u = compose_to_binary(image_u, overlay_u, downsized_alpha)
        composed_v = compose_to_binary(image_v, overlay_v, downsized_alpha)
        composed = composed_y <> composed_u <> composed_v
    #   end
    # })

    {:ok, output} = Image.YUV.new_from_binary(composed, 1920, 1080, :C420)
    Image.write!(output, "output.jpeg")
  end

  # creates a new overlay image with transparent background matching the 
  # base image that way we can merge on top base image without positioning
  # it again
  def resized_and_positioned_overlay(overlay, opts) do
    transparent_background =
      Operation.black!(opts[:width], opts[:height], bands: 4)
      |> Operation.copy!(interpretation: :VIPS_INTERPRETATION_sRGB)

    overlay =
      Operation.composite2!(transparent_background, overlay, :VIPS_BLEND_MODE_OVER,
        x: opts[:x],
        y: opts[:y]
      )

    path = "resized_overlay.png"
    :ok = Vimage.write_to_file(overlay, path)
    path
  end

  defp open_planes(yuv, width, height) do
    half_width = div(width, 2)
    half_height = div(height, 2)
    y_size = width * height
    uv_size = half_width * half_height
    <<y::binary-size(y_size), u::binary-size(uv_size), v::binary-size(uv_size)>> = yuv
    {:ok, y} = Vimage.new_from_binary(y, width, height, 1, :VIPS_FORMAT_UCHAR)
    {:ok, u} = Vimage.new_from_binary(u, half_width, half_height, 1, :VIPS_FORMAT_UCHAR)
    {:ok, v} = Vimage.new_from_binary(v, half_width, half_height, 1, :VIPS_FORMAT_UCHAR)
    {y, u, v}
  end

  defp compose_to_binary(image, overlay, alpha) do
    {:ok, binary} = Operation.ifthenelse!(alpha, overlay, image) |> Vimage.write_to_binary()
    binary
  end
end

Overlay.run()
```

![overlay](https://github.com/akash-akya/vix/assets/9396022/e707d4f7-2268-411d-bc16-01d62418b366)

**Reply by mat-hek** [reply]
> Adjusting it to use alpha channel can be done by simply joining alpha band we get from the overlay png image.

Perfect, thanks!

>  I tried another approach using ifthenelse instead of compose to overlay images, and it is lot faster

Cool! Do I understand correctly that it overlays each pixel when its `alpha > 0`?

> Performs best if the overlay is at the top-left and worse if it is at bottom-right.

I guess we could speed up the bottom case by manually cutting off the top `n` rows, which is just binary split and then concatenation. This `ifthenelse` could have some `offset` option 🤔

**Reply by mat-hek** [reply]
> Hopefully I can get to a Image.YUV.write_to_buffer/2 that becomes a NIF call to libheif

@kipcole9 This lib looks promising, looking forward to it! Operating on RGB would surely be more seamless than on YUV. In case it doesn't work out, we can figure out how to improve the experience of YUV in Image and add YUVa support if you will.
